[Benjamin Sim] iP

Benjamin-Sim opened the pull request:
### DukePro
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [x] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

---

``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Benjamin-Sim committed:
Add greet, echo and exit functionality

Benjamin-Sim committed:
Modify Scanner object usage

Benjamin-Sim committed:
Add list and add text functionality

Benjamin-Sim committed:
Add Task class and mark as done and undone functionality

Benjamin-Sim committed:
Add ToDo, Deadline and Event task types
Add test case

Benjamin-Sim committed:
Add input error handling for invalid commands and adding ToDo, Deadline and Event tasks

Benjamin-Sim committed:
Add delete task functionality

Benjamin-Sim committed:
Add save to hard disk functionality

Benjamin-Sim committed:
Add date and time functionality for Deadline tasks

Benjamin-Sim committed:
Merge branch 'branch-Level-7'

Benjamin-Sim committed:
Merge branch 'branch-Level-8'

parth-io commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,72 @@
+        // dateAndTime is of the form "yyyy-mm-dd hh:mm"
+        // Therefore, the first token is the date and the second token is the time
+
+        // Split string around whitespaces
```
I feel this comment is a bit redundant, as the code is self-explanatory

Benjamin-Sim replied:
I write these comments to help myself easily remember the logic for my code. 😅
But good to know that it's self-explanatory for others

parth-io commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,72 @@
+
+
+
+    
```
These spaces can be deleted

parth-io commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,340 @@
+    public Duke() {
+        // TaskList is initialized in the initialize() method
+    }
+
```
This can be safely deleted to make it less confusing. The default constructor is an empty constructor

parth-io commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,340 @@
+    }
+
+
+    public boolean isValidEventCommand(String[] commands) {
```
The naming for this method can be `hasValidEventCommand`, as the current name reads more like one for a boolean variable

parth-io commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,35 @@
+    }
+    
+
+    @Override
```
I like the `@Override` annonations, makes it clear!

parth-io commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,340 @@
+    public Task createTask(String[] commands) throws Exception {
+        
+        String description = "";
+        boolean isDeadline = false;
```
Might be good to make these private

Benjamin-Sim replied:
I think the accessibility modifier doesn't matter for these variables since they are part of a method, not an instance or class attribute

parth-io reviewed (commented):
In general, I found the code very well structured and easy to follow. The use of fallthroughs in switch statements seems to go against recommended practices, but I don't think it is an issue as the codebase is still small.

Comments can also be deleted; there are many redundant comments. For example,
```java
  // Constructors
    public ToDo(String description) {
        super(description);
    }
```

I found the naming to be very good, and there were long, descriptive names where appropriate, so that was nice and easy to follow code logic.

Benjamin-Sim committed:
Add Gradle support

Benjamin-Sim committed:
Refactor code into more classes for better OOP design

Benjamin-Sim committed:
Reorganize files into packages

Benjamin-Sim committed:
Add JUnit tests for Deadline and Todo classes

Benjamin-Sim committed:
Package app as a JAR file

Benjamin-Sim committed:
Add JavaDoc comments to non-private classes and methods

Benjamin-Sim committed:
Tweak code to comply with coding standard

Benjamin-Sim committed:
Add find task functionality

Benjamin-Sim committed:
Merge branch 'branch-A-CodingStandard'

Benjamin-Sim committed:
Merge branch 'branch-A-JavaDoc'

Benjamin-Sim committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/data/TaskList.java
#	src/main/java/duke/ui/Ui.java

Benjamin-Sim committed:
Tweak code to comply with coding standard

Benjamin-Sim committed:
Fix CheckStyle violations

Benjamin-Sim committed:
Add GUI

Benjamin-Sim committed:
Extract UI messages into Ui.java

Benjamin-Sim committed:
Package GUI files

Benjamin-Sim committed:
Merge branch 'branch-Level-10'

Benjamin-Sim committed:
Add assertions to verify assumptions at several places

The Duke.createTask() method may return null due to the default switch
case. This may cause errors in other parts of the code if a null value
is inserted as a Task in the TaskList.

Let's use an assertion to verify that the caller of createTask() doesn't
receive a null value by accident.

When removing a Task in the Duke.deleteTask() method, we do not check
if a null value is returned by the TaskList.removeTask() method. This
may cause an error when generating the corresponding UI messages if a
null value is used.

Let's use an assertion to verify that the returned Task object is not
null.

The String argument of the Duke.getResponse() method may be null due to
some error in the GUI classes/code. This will cause errors in other
parts of the code if a null value is used as user input.

Let's use an assertion to verify that the given String argument is not
null.

Benjamin-Sim committed:
Refactor the Duke.createTask() method by extracting out the code to
create each type of task into separate helper methods.

The createTask() method is too long due to the code to create each type
of task. This makes the entire method hard to read and debug.

Let's extract out the code to create each type of task (Todo, Deadline
and Event) into separate helper methods. This greatly shortens the
createTask() method, making it easier to read. Furthermore, future
changes in how a task is created can now be applied to the specific
type of task instead of affecting the entire createTask() method.

Benjamin-Sim committed:
Modify the TaskList.searchTasks() method to use streams to perform
the search operation

Benjamin-Sim committed:
Merge pull request #1 from Benjamin-Sim/branch-A-CodeQuality

Merge branch-A-CodeQuality with Master

Benjamin-Sim committed:
Merge branch 'master' into branch-A-Assertions

Benjamin-Sim committed:
Merge branch 'master' into branch-A-Streams

Benjamin-Sim committed:
Merge pull request #2 from Benjamin-Sim/branch-A-Assertions

Merge branch-A-Assertions with Master

Benjamin-Sim committed:
Merge branch 'master' into branch-A-Streams

Benjamin-Sim committed:
Merge pull request #3 from Benjamin-Sim/branch-A-Streams

Merge branch-A-Streams with Master

Benjamin-Sim committed:
Delete dead code in TaskList.java

Benjamin-Sim committed:
Create gradle.yml

Benjamin-Sim committed:
Update gradle.yml

Benjamin-Sim committed:
Update gradlew permissions

Benjamin-Sim committed:
Update gradle.yml

Benjamin-Sim committed:
Fix Checkstyle violations

Benjamin-Sim committed:
Add sort tasks functionality

Benjamin-Sim committed:
Merge branch 'branch-C-Sort'

Benjamin-Sim committed:
Add user guide to README.md

Benjamin-Sim committed:
Add screenshot and fix Checkstyle violations

Benjamin-Sim committed:
Delete Ui.png

Benjamin-Sim committed:
Add Ui.png

Benjamin-Sim committed:
Modify Duke.java to fix bugs
